draft-ietf-httpbis-http2-secondary-certs.md

```diff
@@ -374,47 +369,54 @@ contains the new origin as well, and if so, reuse the connection.
 
 If the TLS certificate does not contain the new origin, but the server has
 claimed support for that origin (with an ORIGIN frame, see
-[I-D.ietf-httpbis-origin-frame]) and advertised support for HTTP-layer
-certificates (see {{setting}}), it MAY send a `CERTIFICATE_NEEDED` frame on the
-stream it will use to make the request. (If the request parameters have not
-already been made available using a `CERTIFICATE_REQUEST` frame, the client will
-need to send the `CERTIFICATE_REQUEST` in order to generate the
-`CERTIFICATE_NEEDED` frame.) The stream represents a pending request to that
-origin which is blocked until a valid certificate is processed.
-
-The request is blocked until the server has responded with a
-`USE_CERTIFICATE` frame pointing to a certificate for that origin. If
-the certificate is already available, the server SHOULD immediately
-respond with the appropriate `USE_CERTIFICATE` frame. (If the
-certificate has not already been transmitted, the server will need to
-make the certificate available as described in {{cert-available}} before
-completing the exchange.)
-
-If the server does not have the desired certificate, it MUST respond with an
-empty `USE_CERTIFICATE` frame. In this case, or if the server has not advertised
-support for HTTP-layer certificates, the client MUST NOT send any requests for
-resources in that origin on the current connection.
+[I-D.ietf-httpbis-origin-frame], or by pushing content) and advertised support
+for HTTP-layer certificates (see {{setting}}), it MAY send a
+`CERTIFICATE_REQUEST` frame describing the desired certificate. Clients MUST NOT
+send requests for the specified origin on the connection until the requested
+certificate has been provided.
 
 ~~~
 Client                                      Server
    <----------------------- (stream 0) ORIGIN --
    -- (stream 0) CERTIFICATE_REQUEST ---------->
    ...
-   -- (stream N) CERTIFICATE_NEEDED ----------->
    <------------------ (stream 0) CERTIFICATE --
-   <-------------- (stream N) USE_CERTIFICATE --
    -- (stream N) GET /from-new-origin --------->
    <----------------------- (stream N) 200 OK --
-
 ~~~
 {: #ex-http2-server-requested title="Client-Requested Certificate"}
 
-Likewise, on each stream where certificate authentication is required,
-the server sends a `CERTIFICATE_NEEDED` frame, which the client
-answers with a `USE_CERTIFICATE` frame indicating the certificate to
-use. If the request parameters or the responding certificate are not
-already available, they will need to be sent as described in
-{{cert-available}} as part of this exchange.
+A server SHOULD provide certificates for an origin before pushing resources from
+it or supplying content referencing the origin. If a client receives a
+`PUSH_PROMISE` referencing an origin for which it has not yet received the
+server's certificate, the client SHOULD verify the server's possession of an
```

Comment:
You use a lot of "SHOULD" here, and I think that they are for the wrong reason.

In this case, you are using SHOULD to avoid having to say *why* the client might want to verify that the server has a valid certificate.  Here, you can say that the verification is mandatory.  In many other cases, you can describe what goal the endpoint might have and how it can achieve that goal.